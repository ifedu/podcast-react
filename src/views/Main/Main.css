.Main {
  border-top: 1px solid #ccc;
}

/* filter */
.Main .filter {
  align-items: center;
  display: flex;
  justify-content: right;
  margin: 16px;
}

.Main .filter span {
  background-color: #2777bf;
  border-radius: 5px;
  color: white;
  margin-right: 8px;
  padding: 0px 4px;
}

.Main .filter input {
  border-radius: 4px;
  border: 1px solid #aaa;
  min-width: 120px;
  max-width: 200px;
  padding: 8px;
  width: 200px;
}

.Main .filter input::placeholder {
  color: #aaa;
}

/* songs */
.Main .cols {
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(4, 1fr);
}

.Main .col {
  cursor: pointer;
  display: flex;
  justify-content: space-around;
  margin-bottom: 100px;
}

.Main .song {
  height: 200px;
  position: relative;
  width: 100%;
}

.Main .background {
  box-shadow: 0 1px 2px 2px #ddd;
  height: 70%;
  position: absolute;
  top: 30%;
  width: 100%;
}

.Main .image {
  display: flex;
  justify-content: center;
  position: relative;
}

.Main .image img {
  border-radius: 100px;
  height: 120px;
  width: 120px;
}

.Main .texts {
  padding: 12px;
  text-align: center;
}

.Main .title {
  font-size: 14px;
  font-weight: bold;
  margin-bottom: 8px;
  max-height: 40px;
  overflow: hidden;
  text-transform: uppercase;
}

.Main .author {
  color: #777;
  font-size: 12px;
}
