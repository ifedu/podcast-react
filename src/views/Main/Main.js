import React from 'react';
import { Link } from 'react-router-dom';

import './Main.css';
import { useLoading } from '../../hooks/LoadingContext';
import { useSongs } from '../../hooks/useSongs';

export function Main() {
  const { setIsLoading } = useLoading();
  const [filterValue, setFilter] = React.useState('');

  const songs = useSongs(setIsLoading);

  function onFilterChange(e) {
    setFilter(e.target.value);
  }

  if (!songs) {
    return null;
  }

  return (
    <div className='Main' data-testid='main'>
      <div className='filter'>
        <span>{songs?.entry?.length}</span>
        <input placeholder='Filter podcasts...' onChange={onFilterChange}/>
      </div>

      <div className='cols'>
        {
          songs?.entry.map((song, i) => {
            if (
              filterValue === '' ||
              song['im:name'].label.toUpperCase().includes(filterValue.toUpperCase()) ||
              song['im:artist'].label.toUpperCase().includes(filterValue.toUpperCase())
            ) {
              return <Link key={i} to={`/podcast/${song.id.attributes['im:id']}`} state={{ song }}  data-testid='link-to-podcast'>
                <div className='col'>
                  <div className='song'>
                    <div className='background'></div>

                    <div className='image'>
                      <img src={song['im:image'][2].label} alt='All Songs Considered'/>
                    </div>

                    <div className='texts'>
                      <div className='title'>{song['im:name'].label}</div>
                      <div className='author'>Author: {song['im:artist'].label}</div>
                    </div>
                  </div>
                </div>
              </Link>
            }

            return null;
          })
        }
      </div>
    </div>
  );
}
