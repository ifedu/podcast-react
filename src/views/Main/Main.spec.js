import { cleanup, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { create } from 'react-test-renderer' ;

import { Main } from './Main';

jest.mock('../../hooks/LoadingContext.js', () => ({
  useLoading: () => ({
    setIsLoading: jest.fn(),
  }),
}));

jest.mock('../../hooks/useSongs.js', () => ({
  useSongs: () => ({
    entry: [
      {
        'im:artist': {},
        'im:image': [{}, {}, {}],
        'im:name': {},
        id: {
          attributes: {
            'im:id': 1,
          },
        },
      },
    ]
  }),
}));

const getContent = (pathname = '/') => {
  return <MemoryRouter initialEntries={[{ pathname }]}>
    <Routes>
      <Route path='/' element={
        <Main/>
      }></Route>

      <Route path='/podcast/1/' element={
        <div data-testid='podcast'></div>
      }></Route>
    </Routes>
  </MemoryRouter>;
}

describe('Podcast', () => {
  const view = getContent();

  afterEach(cleanup);

  it('renders correctly', () => {
    const tree = create(view).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the main view', async() => {
    render(view);

    await waitFor(() => {
      const main = screen.queryByTestId('main');
      expect(main).toBeInTheDocument();
    });
  });

  it('navigates from main view to the podcast view', async () => {
    render(view);

    await waitFor(() => {
      const main = screen.queryByTestId('main');
      expect(main).toBeInTheDocument();
    });

    let link = screen.queryByTestId('link-to-podcast');
    userEvent.click(link);

    await waitFor(() => {
      const episode = screen.queryByTestId('podcast');
      expect(episode).toBeInTheDocument();
    });
  });
});
