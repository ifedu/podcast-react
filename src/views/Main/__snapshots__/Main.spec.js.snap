// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Podcast renders correctly 1`] = `
<div
  className="Main"
  data-testid="main"
>
  <div
    className="filter"
  >
    <span>
      1
    </span>
    <input
      onChange={[Function]}
      placeholder="Filter podcasts..."
    />
  </div>
  <div
    className="cols"
  >
    <a
      data-testid="link-to-podcast"
      href="/podcast/1"
      onClick={[Function]}
    >
      <div
        className="col"
      >
        <div
          className="song"
        >
          <div
            className="background"
          />
          <div
            className="image"
          >
            <img
              alt="All Songs Considered"
            />
          </div>
          <div
            className="texts"
          >
            <div
              className="title"
            />
            <div
              className="author"
            >
              Author: 
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
